types/lodash/common/function.d.ts

```diff
@@ -368,18 +368,16 @@ declare module "../index" {
          */
         trailing?: boolean | undefined;
     }
-    interface DebouncedFunc<T extends (...args: any[]) => any> {
-        /**
-         * Call the original function, but applying the debounce rules.
-         *
-         * If the debounced function can be run immediately, this calls it and returns its return
-         * value.
-         *
-         * Otherwise, it returns the return value of the last invocation, or undefined if the debounced
-         * function was not invoked yet.
-         */
-        (...args: Parameters<T>): ReturnType<T> | undefined;
-
+    /**
+     * Call the original function, but applying the debounce rules.
+     *
+     * If the debounced function can be run immediately, this calls it and returns its return
+     * value.
+     *
+     * Otherwise, it returns the return value of the last invocation, or undefined if the debounced
+     * function was not invoked yet.
+     */
+    type DebouncedFunc<T extends (...args: any[]) => any> = (T | (() => undefined)) & {
```

Comment:
I'm having a hard time understanding if that `| (() => undefined)` part of the type is right. The previous type implied we could not call a `DebouncedFunc<T>` with 0 arguments (i.e. had signature `(...args: Parameters<T>): ReturnType<T> | undefined`), whereas this new type implies we can call it with 0 arguments and get `undefined` as result. Could you explain that?

Reply:
I think you are confusing the concept of function overloading (which should be done with intersection `&`, not union `|`) with the union of callable functions. My solution may be a bit hard to understand because I am using the union of callable types. I will try to explain how it works in the following steps:

1. Function overloading can be done using intersection
```ts
// With interface overload
interface CallableOverloadedWithInterface {
    (a: string): 0;
    (b: number): 1;
};

declare const callableFnUsingInterface: CallableOverloadedWithInterface;
callableFnUsingInterface('string'); // $ExpectedType 0
callableFnUsingInterface(42); // $ExpectedType 1

// With intersection overload
type CallableOverloadedWithIntersection =
    & ((a: string) => 0)
    & ((a: number) => 1);

declare const callableFnUsingIntersection: CallableOverloadedWithIntersection;
callableFnUsingIntersection(42); // $ExpectedType 1
```
So you can see that `CallableOverloadedWithInterface` has exactly the same type signature as ` CallableOverloadedWithIntersection`.

2. Callable unions act differently.
Typescript@3.3 introduces improved behavior for calling union types.
You can see notes here: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-3.html#improved-behavior-for-calling-union-types
> "In TypeScript 3.3, the parameters of these signatures are intersected together to create a new signature."
So basically when we write code like this:
```ts
type callableUnion = (a: string) => 0 | (a: string | undefined) => 1;
// `a` parameter will be reduced to: `(string) & (string | undefined)` -> `string`
// the type returned by `callableUnion` will be the union of the types returned from all signatures
declare const fn: callableUnion; // (a: string) => 0 | 1;
```
we can observe that the return value is the union of each signature return value.

3. Using this technique, we can extend the type returned by the function!
`(T | (() => undefined))` will add undefined to `T` return type.
And that's how my solution works. It adds undefined to the type returned by the original function.
```ts
// Debounce type implementation
type DebouncedFn<T extends (...args: any[]) => any> = (T | (() => undefined)) & {
    cancel(): void;
    flush(): ReturnType<T> | undefined;
};
declare function debounce<T extends (...args: any[]) => any>(f: T): DebouncedFn<T>;

// Example fn and it's debounce
const exampleFn = function<T>(a: T, b: number): T {
    return a;
};
const debouncedFn = debounce(exampleFn);

// @ts-expect-error
debouncedFn(); // Expected 2 arguments, but got 0.
debouncedFn('asd', 42); // "asd" | undefined
```
This ensures that the generics are preserved and the types are correct.

### Quick Q&A based on my research:

1. Since the parameters of these signatures are intersected together, why `((a: string) => void) | () => void)` first parameter is string, not never? (`string & undefined` should be `never`)
The answer can be found in the source code of typescript complier (or in one of ts@3.3 pull requests):
https://github.com/microsoft/TypeScript/pull/29011/files#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8R6873-R6874
```ts
// src/compiler/checker.ts 
/* 6871: */ for (let i = 0; i < longestCount; i++) {
/* 6872: */    const longestParamType = tryGetTypeAtPosition(longest, i)!;
/* 6873: */    const shorterParamType = tryGetTypeAtPosition(shorter, i) || unknownType;
/* 6874: */    const unionParamType = getIntersectionType([longestParamType, shorterParamType]);
```
We can see that in the case of absence of a parameter, it gets the default unknown type. (`string & unknown` === `string`)

2. Why `((a: string) => void) | ((a: any) => void)` is equal to `(a: string) => void`, not `(a: any) => void`?
Typescript under the hood uses `TypeFlags.AnyOrUnknown` / `TypeFlags.Wildcard` in these cases.
Unfortunately, I'm not sure exactly where in the code this happens.

Hope this will dispel your doubts :)

Reply:
This is a great explanation of your intent, and it does seem to work. However, I’m very nervous about this change because union calls are complicated, they have limitations, and `lodash.debounce` is one of the most widely used functions from any JS library ever. Because of that, I think we may want you to do a little extra proof that this does not break projects in the wild. This could take the form of finding a couple big open source dependents, copying your changes into their `node_modules/@types/lodash`, and seeing if you get any (additional) build errors.

As an aside, did you consider something like this (using your playground link from the description)?

```ts
// My proposal
declare function new_debounce<ParamsT extends any[], ReturnT>(func: (...args: ParamsT) => ReturnT): new_DebouncedFunc<ParamsT, ReturnT>;
type new_DebouncedFunc<ParamsT extends any[], ReturnT> = (...args: ParamsT) => ReturnT | undefined;
```

I think that might capture your intent a little better, and doesn’t rely on union calls.

Reply:
After I found a bug in the union calls recently, I have to admit you are right - union calls are unstable at this moment. The code that you proposed solves the problem presented by me, and moreover, is more readable. I will make changes in a few days

Reply:
Unless a bunch of package owners show up and approve this, I think we’ll still want to see some proof of this working in the wild. These sorts of changes that look great in theory have a bad habit of backfiring in unexpected ways, so the confidence bar for changes labeled "Critical Package" is _very_ high.